### PR TITLE
Update FindUBUsingLatticeParameters for small numbers of peaks

### LINKED ISSUE #18039: Additional options for FindUBUsingLatticeParameters
A request from Pascal. He has data where he knows what the lattice parameters should be but frequently finds that this algorithm fails because of not having enough visible peaks in his data to successfully index > 3 within a small tolerance from which to optimise the UB. 

This frequently causes the algorithm to fail because it could not successfully index the >3 peaks needed for optimisation. For practical purposes he'd rather get any UB than get a perfect UB.

He also mentioned it would be useful to have an option to be able to optimise a,b, and c but not alpha, beta, or gamma.

I would suggest two new boolean properties:
 - Fix angles - this will find a UB then run ```Optimise_UB``` as normal but only on the first a,b,c, and volume parameters.
 - Fix all - this will find a UB but will not run ```Optimise_UB``` afterwards, avoiding the error.

In both cases I think we should output a warning about what is happening.

### PR DESCRIPTION
**Pull request by samueljackson92**
Pascal has been having problems with this algorithm when using a very small number of peaks. Technically we should be able to use as few as 3 peaks but found the algorithm often failed after UB optimisation. Here we do two things:

1. Remove the additional UB optimisation in the `FindUBUsingLatticeParameters` algorithm. This is already performed in the  `Find_UB`, but in try catch. Now if no improvement is found, a UB is still returned.
2. A new option called `FixParameters` (default false) has been added as requested by Pascal. When this option is used the algorithm will find a UB but will not attempt to optimise using indexed HKL and Q values.

**To test:**
1. Check the tests pass
2. Try to find a UB on some data. I've provided two peaks workspaces ([peaks_data.zip](https://github.com/mantidproject/mantid/files/610891/peaks_data.zip)). One with 3 peaks and one with 4. You can use the following script to try out the changes:

```python
import os.path
path = '/Users/samueljackson/Desktop/'
peaks = LoadIsawPeaks(Filename=os.path.join(path, 'WISH00035920.peaks'))
FindUBUsingLatticeParameters(PeaksWorkspace=peaks, a=3.89, b=3.89, c=20.69, alpha=90, beta=90, gamma=90)
FindUBUsingLatticeParameters(PeaksWorkspace=peaks, a=3.89, b=3.89, c=20.69, alpha=90, beta=90, gamma=90, FixParameters=True)

peaks2 = LoadIsawPeaks(Filename=os.path.join(path, 'WISH00035920_2.peaks'))
FindUBUsingLatticeParameters(PeaksWorkspace=peaks2, a=3.89, b=3.89, c=20.69, alpha=90, beta=90, gamma=90)
FindUBUsingLatticeParameters(PeaksWorkspace=peaks2, a=3.89, b=3.89, c=20.69, alpha=90, beta=90, gamma=90, FixParameters=True)
````
You should see output listed similar to the below. Note that all of these return a UB despite the low number of peaks (in the case of the first workspace only 2 out of three are indexed).

```
LoadIsawPeaks started
LoadIsawPeaks successful, Duration 0.50 seconds
FindUBUsingLatticeParameters started
Error = 0.0176354
UB = Matrix(3,3)0.0573513,0.198464,-0.0287652,0.244488,-0.00988809,0.0148194,0.0549668,-0.163092,-0.0359024
New UB will index 2 Peaks out of 3 with tolerance 0.150
Lattice Parameters:    3.890000    3.890000   20.690000   90.000000   90.000000   90.000000   313.083149
Lattice Parameters (Refined - Input):   -0.000000   -0.000000    0.000000    0.000000    0.000000    0.000000
FindUBUsingLatticeParameters successful, Duration 1.10 seconds
FindUBUsingLatticeParameters started
Error = 0.0176354
UB = Matrix(3,3)0.0573513,0.198464,-0.0287652,0.244488,-0.00988809,0.0148194,0.0549668,-0.163092,-0.0359024
New UB will index 2 Peaks out of 3 with tolerance 0.150
Lattice Parameters:    3.890000    3.890000   20.690000   90.000000   90.000000   90.000000   313.083149
Lattice Parameters (Refined - Input):   -0.000000   -0.000000    0.000000    0.000000    0.000000    0.000000
FindUBUsingLatticeParameters successful, Duration 1.14 seconds
LoadIsawPeaks started
LoadIsawPeaks successful, Duration 0.54 seconds
FindUBUsingLatticeParameters started
Error = 3.74849e-30
UB = Matrix(3,3)0.177645,0.219583,0.000918923,0.0515916,0.00840966,0.0499153,0.244531,-0.138228,0.0010567
New UB will index 3 Peaks out of 4 with tolerance 0.150
Lattice Parameters:    3.332754    3.862780   20.426618   91.336109  101.171713   93.725987   257.264496
Lattice Parameters (Refined - Input):   -0.557246   -0.027220   -0.263382    1.336109   11.171713    3.725987
FindUBUsingLatticeParameters successful, Duration 1.11 seconds
FindUBUsingLatticeParameters started
Error = 0.0261459
UB = Matrix(3,3)0.141237,0.212968,-0.00525575,0.0402145,0.00664644,0.0477211,0.210997,-0.143823,-0.00557722
New UB will index 3 Peaks out of 4 with tolerance 0.150
Lattice Parameters:    3.890000    3.890000   20.690000   90.000000   90.000000   90.000000   313.083149
Lattice Parameters (Refined - Input):    0.000000    0.000000    0.000000    0.000000    0.000000    0.000000
FindUBUsingLatticeParameters successful, Duration 1.18 seconds
```

Fixes #18039 .

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?

- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Crystal/src/FindUBUsingLatticeParameters.cpp [hunk 1]
```diff
@@ -91,18 +95,6 @@ void FindUBUsingLatticeParameters::exec() {
...
-    std::vector<V3D> miller_ind;
-    std::vector<V3D> indexed_qs;
-    double fit_error;
-    miller_ind.reserve(q_vectors.size());
-    indexed_qs.reserve(q_vectors.size());
-    IndexingUtils::GetIndexedPeaks(UB, q_vectors, tolerance, miller_ind,
-                                   indexed_qs, fit_error);
-
-    IndexingUtils::Optimize_UB(UB, miller_ind, indexed_qs, sigabc);
-
```
Comment (raquelalvarezbanos) [inline Framework/Crystal/src/FindUBUsingLatticeParameters.cpp]: I guess this is the block you've removed that you mention in the description. Does it mean that this was being done twice?

File: Framework/Crystal/src/FindUBUsingLatticeParameters.cpp [hunk 2]
```diff
@@ -44,6 +44,8 @@ void FindUBUsingLatticeParameters::init() {
                         "Lattice parameter gamma");
   this->declareProperty("NumInitial", 15, moreThan2Int,
                         "Number of Peaks to Use on First Pass(15)");
+  this->declareProperty("FixParameters", false,
+                        "Do not optimise the UB after finding the orientation");
```
Comment (samueljackson92) [inline Framework/Crystal/src/FindUBUsingLatticeParameters.cpp]: @raquelalvarezbanos Noted. I shall change the property order.

File: Framework/Crystal/src/FindUBUsingLatticeParameters.cpp (hunk 1 above)
Comment (samueljackson92) [inline Framework/Crystal/src/FindUBUsingLatticeParameters.cpp]: @raquelalvarezbanos Yes. I believe that this was effectively called twice, once in `Find_UB` towards the end of the method and once in `FindUBUsingLatticeParameters`. I don't see why It needed to be called twice other than to get the errors for the lattice parameters, which is unnecessary now we pass the lattice as an argument. Removing it did not break any unit tests, which gave me some confidence that it is redundant. Perhaps someone with more knowledge of this code ( @AndreiSavici ?) could verify that second call is unnecessary.

**Comment by raquelalvarezbanos** [comment]
Code changes look good to me, but as @samueljackson92 says, it would be good if a second reviewer more familiar with the code had a second look.

:shipit:

**Comment by VickieLynch** [comment]
I compared with nightly using TOPAZ data and the parameter errors are the same, so the deleted  call to Optimize_UB was unnecessary.  :shipit:

File: Framework/Geometry/src/Crystal/IndexingUtils.cpp [hunk 3]
```diff
@@ -53,12 +52,11 @@ const constexpr double RAD_TO_DEG = 180. / M_PI;
   @param  q_vectors           std::vector of V3D objects that contains the
                               list of q_vectors that are to be indexed
                               NOTE: There must be at least 2 q_vectors.
-  @param  a                   First unit cell edge length in Angstroms.
-  @param  b                   Second unit cell edge length in Angstroms.
-  @param  c                   Third unit cell edge length in Angstroms.
-  @param  alpha               First unit cell angle in degrees.
-  @param  beta                second unit cell angle in degrees.
-  @param  gamma               third unit cell angle in degrees.
+  @param  lattice							The
```
Comment (AndreiSavici) [inline Framework/Geometry/src/Crystal/IndexingUtils.cpp]: Can you fix the text alignment?

File: Framework/Geometry/src/Crystal/IndexingUtils.cpp [hunk 4]
```diff
@@ -880,12 +887,8 @@ double IndexingUtils::Optimize_Direction(V3D &best_vec,
     @param UB                 This will be set to the UB matrix that best
                               indexes the supplied list of q_vectors.
     @param q_vectors          List of locations of peaks in "Q".
-    @param a                  Lattice parameter "a".
-    @param b                  Lattice parameter "b".
-    @param c                  Lattice parameter "c".
-    @param alpha              Lattice parameter alpha.
-    @param beta               Lattice parameter beta.
-    @param gamma              Lattice parameter gamma.
+    @param cell		      Unit cell defining the parameters a,b,c and
```
Comment (AndreiSavici) [inline Framework/Geometry/src/Crystal/IndexingUtils.cpp]: Fix spacings

File: Framework/Geometry/src/Crystal/IndexingUtils.cpp (hunk 3 above)
Comment (samueljackson92) [inline Framework/Geometry/src/Crystal/IndexingUtils.cpp]: Agreed. Annoying it seems that Qt creator didn't convert tabs to spaces properly here.

**Comment by samueljackson92** [comment]
@VickieLynch Thanks for checking that. Good to know.

**Comment by AndreiSavici** [comment]
Builds failed for unrelated reasons

**Pull request merged**